### PR TITLE
Catalog Manager Test 1

### PR DESCRIPTION
**Pull request by Moliholy**
Makes the AbstractCatalogManager templated. It also defines the new MockCatalogManager class.

### CODE REVIEW

File: cvmfs/catalog_mgr.h [hunk 1]
```diff
@@ -115,16 +115,18 @@ class InodeGenerationAnnotation : public InodeAnnotation {
...
-
+template <class CatalogT>
 class AbstractCatalogManager;
+
 /**
  * Here, the Cwd Buffer is registered in order to save the inodes of
  * processes' cwd before a new catalog snapshot is applied
  */
+template <class CatalogT>
 class RemountListener {
```
Comment (jblomer) [inline cvmfs/catalog_mgr.h]: Actually I think that we don't use the `RemountListener` anywhere anymore.  Can you try to remove it, including the calling code in `AbstractCatalogManager`?

File: cvmfs/catalog_mgr.h [hunk 2]
```diff
@@ -142,8 +144,12 @@ class RemountListener {
  *   catalog_manager->Init();
  *   catalog_manager->Lookup(<inode>, &<result_entry>);
  */
+template <class CatalogT>
 class AbstractCatalogManager : public SingleCopy {
  public:
+  typedef std::vector<CatalogT*> CatalogList;
+
+ public:
```
Comment (jblomer) [inline cvmfs/catalog_mgr.h]: Can be removed

File: cvmfs/catalog_mgr.h [hunk 3]
```diff
@@ -142,8 +144,12 @@ class RemountListener {
  *   catalog_manager->Init();
  *   catalog_manager->Lookup(<inode>, &<result_entry>);
  */
+template <class CatalogT>
```
Comment (jblomer) [inline cvmfs/catalog_mgr.h]: It would be good to add a line of documentation here: `CatalogT` is either `Catalog` or the mocked catalog (or perhaps at some point a read-writable catalog).

File: cvmfs/catalog_mgr_impl.h [hunk 4]
```diff
@@ -1,3 +1,13 @@
+/*
+ * catalog_mgr_impl.h
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: This comment block should be removed (it's autogenerated by the editor, right?)

File: cvmfs/catalog_mgr_impl.h [hunk 5]
```diff
@@ -1,3 +1,13 @@
...
+ *
+ *  Created on: Aug 13, 2015
+ *      Author: molina
+ */
+
+#ifndef CVMFS_CATALOG_MGR_IMPL_H_
+#define CVMFS_CATALOG_MGR_IMPL_H_
+
 /**
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Should be the first comment block

Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: Very true. This is fixed in the pull request number 3, where I have the cpplint fixes

File: cvmfs/catalog_mgr_impl.h (hunk 4 above)
Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: Fixed in the pull request number 3

File: cvmfs/catalog_mgr_impl.h [hunk 6]
```diff
@@ -533,13 +554,14 @@ Catalog* AbstractCatalogManager::FindCatalog(const PathString &path) const {
  * @param attached_catalog is set to the searched catalog, if not NULL
  * @return true if catalog is already present, false otherwise
  */
-bool AbstractCatalogManager::IsAttached(const PathString &root_path,
-                                        Catalog **attached_catalog) const
+template <class CatalogT>
+bool AbstractCatalogManager<CatalogT>::IsAttached(const PathString &root_path,
+                                        CatalogT **attached_catalog) const
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: That's a strange indentation

File: cvmfs/catalog_mgr_impl.h [hunk 7]
```diff
@@ -554,13 +576,14 @@ bool AbstractCatalogManager::IsAttached(const PathString &root_path,
  * nested catalog for the given path.
  * The final leaf nested catalog is returned.
  */
-bool AbstractCatalogManager::MountSubtree(const PathString &path,
-                                          const Catalog *entry_point,
-                                          Catalog **leaf_catalog)
+template <class CatalogT>
+bool AbstractCatalogManager<CatalogT>::MountSubtree(const PathString &path,
+                                          const CatalogT *entry_point,
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Also strange indentation.  How about

```c
bool AbstractCatalogManager<CatalogT>::MountSubtree(
  const CatalogT *entry_point,
  ...
```

File: cvmfs/catalog_mgr_impl.h [hunk 8]
```diff
@@ -605,11 +628,12 @@ bool AbstractCatalogManager::MountSubtree(const PathString &path,
  * Load a catalog file and attach it to the tree of Catalog objects.
  * Loading of catalogs is implemented by derived classes.
  */
-Catalog *AbstractCatalogManager::MountCatalog(const PathString &mountpoint,
+template <class CatalogT>
+CatalogT *AbstractCatalogManager<CatalogT>::MountCatalog(const PathString &mountpoint,
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Indentation, see above

File: cvmfs/catalog_mgr_impl.h [hunk 9]
```diff
@@ -643,8 +667,9 @@ Catalog *AbstractCatalogManager::MountCatalog(const PathString &mountpoint,
  * @param new_catalog the catalog to attach to this CatalogManager
  * @return true on success, false otherwise
  */
-bool AbstractCatalogManager::AttachCatalog(const string &db_path,
-                                           Catalog *new_catalog)
+template <class CatalogT>
+bool AbstractCatalogManager<CatalogT>::AttachCatalog(const string &db_path,
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Indentation, see above

File: cvmfs/catalog_mgr_impl.h [hunk 10]
```diff
@@ -749,17 +778,17 @@ string AbstractCatalogManager::PrintHierarchyRecursively(const Catalog *catalog,
     string(catalog->path().GetChars(), catalog->path().GetLength()) + "\n";
 
   CatalogList children = catalog->GetChildren();
-  for (CatalogList::const_iterator i = children.begin(), iEnd = children.end();
-       i != iEnd; ++i)
-  {
+  typename CatalogList::const_iterator i = children.begin();
+  typename CatalogList::const_iterator iend = children.end();
+  for (;i != iend; ++i) {
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Spacing

File: cvmfs/catalog_mgr_impl.h (hunk 8 above)
Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: I'll fix all these in the final pull request (number 3)

File: cvmfs/catalog_mgr_impl.h [hunk 11]
```diff
@@ -769,3 +798,8 @@ void AbstractCatalogManager::EnforceSqliteMemLimit() {
 }
 
 }  // namespace catalog
+
+
+
+
+#endif /* CVMFS_CATALOG_MGR_IMPL_H_ */
```
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Spacing

**Comment by jblomer** [comment]
Actually the changes are more gentle than I initially thought.  It looks quite good!

**Comment by Moliholy** [comment]
All the "cpplint" warnings pointed out in the comments are already fixed in the last version, except the ones with the indentation of the parameters, which are being fixed now.

**Comment by jblomer** [comment]
Ok, out of general paranoia I'll keep this patch set anyway open until after the release (hopefully only a few days more).

**Comment by reneme** [comment]
The patch set looks good to me, too. Actually I was thinking of making the `AbstractCatalogManager` a template since quite a while, since I had good experiences with this code style in the `CatalogTraversal<>` class but needed to play some tricks to make it work with the old-style `AbstractCatalogManager`.

Now with this in place, we could actually refactor the `Catalog` class to use [CRTP](http://www.wikiwand.com/en/Curiously_recurring_template_pattern) like plenty of other classes (`Database`, `AbstractObjectFetcher`, `ConcurrentWorker`, ...) in the system already do. That saves you from all the `virtual` methods and gives you nice type safety.

**Comment by jblomer** [comment]
@Moliholy please let me know once all the comments are addressed.

File: cvmfs/catalog_mgr.h (hunk 1 above)
Comment (Moliholy) [inline cvmfs/catalog_mgr.h]: RemounListener removed

File: cvmfs/catalog_mgr_impl.h (hunk 9 above)
Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: Fixed in this PR

File: cvmfs/catalog_mgr.h (hunk 3 above)
Comment (Moliholy) [inline cvmfs/catalog_mgr.h]: Documentation added in the last commit of this PR

**Comment by Moliholy** [comment]
All comment addressed. Moving to https://github.com/cvmfs/cvmfs/pull/1058

File: cvmfs/catalog_mgr_impl.h (hunk 4 above)
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Please move the commit forward to this PR.  The point of this is to not merge something that has dangling changes.  What gets merged should run correctly through test suite and cpplint and should be consistent and beyond remaining doubts, if humanly possible.

File: cvmfs/catalog_mgr_impl.h (hunk 5 above)
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Please move forward (see above).  Perhaps you got confused about my earlier comment that cpp fixes can be a separate pull request.  That refers to existing problems, i.e. a pull request can contain a number of fixes for the existing code base.  New PRs should be correct before merging.

**Comment by Moliholy** [comment]
Updated including the cpplint warnings and indentation problems.

File: cvmfs/catalog_mgr_impl.h (hunk 5 above)
Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: This comment block is now duplicated and should be removed

Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: It appears correctly in the full changes. I don't know why in the conversation it still appear like in the old version

Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Hm, I don't know what happened.  But if I try to merge, I'd merge with both comment blocks.

Comment (Moliholy) [inline cvmfs/catalog_mgr_impl.h]: That's pretty weird... I'll try to figure it out

Comment (jblomer) [inline cvmfs/catalog_mgr_impl.h]: Never mind, I'll remove it as part of the merging.
